pkg/mutation/mutation_function.go

```diff
@@ -60,108 +111,68 @@ func mutate(m types.Mutator, current interface{}, previous interface{}, depth in
-			if err := mutate(m, next, current, depth+1); err != nil {
-				return err
+			if !s.tester.MissingOkay(depth) {
+				return false, nil, nil
 			}
+			next, err := s.createMissingElement(depth)
+			if err != nil {
+				return false, nil, err
+			}
+			shallowCopy = append(shallowCopy, next)
+			m, _, err := s.mutateInternal(next, depth+1)
+			if err != nil {
+				return false, nil, err
+			}
+			mutated = mutated || m
 		}
+		return mutated, shallowCopy, nil
 	default:
-		return fmt.Errorf("invalid type pathEntry type: %T", pathEntry)
-	}
-	return nil
-}
-
-func addValue(m types.Mutator, current interface{}, previous interface{}, depth int) error {
-	pathEntry := m.Path().Nodes[depth]
-	switch castPathEntry := pathEntry.(type) {
-	case *parser.Object:
-		return setObjectValue(m, castPathEntry, current, depth)
-	case *parser.List:
-		return setListElementToValue(m, current, previous, castPathEntry, depth)
+		return false, nil, fmt.Errorf("invalid type pathEntry type: %T", pathEntry)
 	}
-	return nil
 }
 
-func setObjectValue(m types.Mutator, pathEntry *parser.Object, current interface{}, depth int) error {
-	key := pathEntry.Reference
-	switch m.(type) {
-	case *AssignMetadataMutator:
-		if elementValue, found, err := nestedString(current, key); err != nil {
-			return err
-		} else if found {
-			log.Info("Mutated value already present", "field", key, "value", elementValue)
-			return nil
-		}
-	}
-	value, err := m.Value()
-	if err != nil {
-		return err
-	}
-	if err = setNestedField(current, value, key); err != nil {
-		log.Error(err, "Failed to mutate object", "field", key, "value", value)
-		return err
-	}
-	return nil
-}
-
-func setListElementToValue(m types.Mutator, current interface{}, previous interface{}, listPathEntry *parser.List, depth int) error {
-	currentAsList, ok := current.([]interface{})
-	if !ok {
-		return fmt.Errorf("mismatch between path entry (type: list) and received object (type: %T). Path: %+v", current, listPathEntry)
-	}
+func (s *mutatorState) setListElementToValue(currentAsList []interface{}, listPathEntry *parser.List, depth int) (bool, []interface{}, error) {
 	if listPathEntry.Glob {
-		return fmt.Errorf("last path entry can not be globbed")
+		return false, nil, fmt.Errorf("last path entry can not be globbed")
 	}
-	newValue, err := m.Value()
+	newValue, err := s.mutator.Value()
 	if err != nil {
-		log.Error(err, "error getting mutator value for mutator %+v", m)
-		return err
+		log.Error(err, "error getting mutator value for mutator %+v", s.mutator)
+		return false, nil, err
 	}
 	newValueAsObject, ok := newValue.(map[string]interface{})
 	if !ok {
-		return fmt.Errorf("last path entry of type list requires an object value, pathEntry: %+v", listPathEntry)
+		return false, nil, fmt.Errorf("last path entry of type list requires an object value, pathEntry: %+v", listPathEntry)
 	}
 
 	key := listPathEntry.KeyField
 	keyValue := *listPathEntry.KeyValue
```

Comment:
I think this dereference is safe because we know Glob==false, but that requires a lot of knowledge of the parser. It might be worth adding a KeyValue accessor that is nil-safe just in case.

Reply:
+1

I think the correct answer is to return an error if we get an unexpected `nil` here.

Reply:
Agreed.

Reply:
added nil checking